Fusing stencil kernels for Mobius fermion with immediately followed fifth dimension operators

hummingtree opened the pull request:
This PR is based on #1056, and should be merged after #1056. It was #1130, but since `github` does not allow changing target branch of a PR, here it is.

This PR adds the option to run the even-odd preconditioned `MdagM`, `M`, and `Mdag` operators for Mobius fermion with fused kernels. These fused kernels fuse the Wilson stencil kernel `Dslash4` with immediately followed fifth dimension operators, such as `m5inv`, `m5pre`, `m5mobius`, or any combinations of these.

To run the fused kernels, an option `use_mobius_fused_kernel` (default to `true`) is added to `InvertParam`. A corresponding command line option `--mobius-fused-kernel <true/false>` (default to `true`) is also added to the various tests.

More tests are ongoing, and single GPU performance of the Mobius `MdagM` operators has increased by 20-30% as a result of using the fused kernels.

Closes #193.

A few remaining pieces:
- [x] A scaling plot with and without the fused kernel.
- [x] ~~See how this works with `NVSHMEM`.~~
- [x] And remember to restore the max Ls.

The fused kernel implementation aggressively uses partial-threadblock `__syncthreads()` which is only possible with the _Independent Thread Scheduling_ feature on Volta and later GPUs. This makes it incompatible with the NVSHMEM dslash policies, so currently fused kernels is disabled when NVSHMEM is turned on at compile time.

A few scaling captured on selene.

![single-32](https://user-images.githubusercontent.com/12970356/115567536-5b3ee580-a289-11eb-8e21-8966bdb515da.png)
![single-24](https://user-images.githubusercontent.com/12970356/115567537-5b3ee580-a289-11eb-81f1-36470a9985d0.png)
![half-32](https://user-images.githubusercontent.com/12970356/115567539-5b3ee580-a289-11eb-9c71-5fa3e494eab0.png)
![half-24](https://user-images.githubusercontent.com/12970356/115567541-5bd77c00-a289-11eb-9e3a-1ea0a63e2de4.png)

Weak scaling on Summit captured by @cpviolator.

| Precision-L-Nnodes | non-fused-GFLOPS | fused-GFLOPS | speed up |
| :-: |:-:| :-:| :-: |
|   half-24- 1 |   4742 |   5569 | 1.17 |
|   half-24- 2 |   3478 |   3687 | 1.06 |
|   half-24- 3 |   3541 |   3646 | 1.03 |
|   half-24- 4 |   3609 |   3772 | 1.05 |
|   half-24- 8 |   2387 |   2566 | 1.07 |
|   half-24-16 |   2278 |   2371 | 1.04 |
|   half-24-32 |   2243 |   2482 | 1.11 |
|   half-32- 1 |   4710 |   5522 | 1.17 |
|   half-32- 2 |   3569 |   3784 | 1.06 |
|   half-32- 3 |   3600 |   3775 | 1.05 |
|   half-32- 4 |   3274 |   3560 | 1.09 |
|   half-32- 8 |   2289 |   2546 | 1.11 |
|   half-32-16 |   2260 |   2490 | 1.10 |
|   half-32-32 |   2254 |   2532 | 1.12 |
| single-24- 1 |   2838 |   4058 | 1.43 |
| single-24- 2 |   2064 |   2155 | 1.04 |
| single-24- 3 |   2110 |   2201 | 1.04 |
| single-24- 4 |   2073 |   2181 | 1.05 |
| single-24- 8 |   1380 |   1517 | 1.10 |
| single-24-16 |   1281 |   1369 | 1.07 |
| single-24-32 |   1281 |   1447 | 1.13 |
| single-32- 1 |   2854 |   4014 | 1.41 |
| single-32- 2 |   2117 |   2187 | 1.03 |
| single-32- 3 |   2163 |   2184 | 1.01 |
| single-32- 4 |   1933 |   2106 | 1.09 |
| single-32- 8 |   1355 |   1465 | 1.08 |
| single-32-16 |   1327 |   1451 | 1.09 |
| single-32-32 |   1333 |   1453 | 1.09 |
| double-24- 1 |   1299 |   1781 | 1.37 |
| double-24- 2 |    927 |   1077 | 1.16 |
| double-24- 3 |   1070 |   1097 | 1.03 |
| double-24- 4 |   1054 |   1084 | 1.03 |
| double-24- 8 |    678 |    757 | 1.12 |
| double-24-16 |    644 |    688 | 1.07 |
| double-24-32 |    647 |    718 | 1.11 |
| double-32- 1 |   1307 |   1799 | 1.38 |
| double-32- 2 |   1036 |   1106 | 1.07 |
| double-32- 3 |   1066 |   1101 | 1.03 |
| double-32- 4 |    940 |   1037 | 1.10 |
| double-32- 8 |    690 |    745 | 1.08 |
| double-32-16 |    653 |    718 | 1.10 |
| double-32-32 |    685 |    754 | 1.10 |

hummingtree committed:
Add the fused kernel d4 + m5 instantiations. Note only interior kernels are enabled, and the mas_ls has been reduced to 16.

hummingtree committed:
Add code that does the fusion of m5inv and m5pre. Change the variable names for the m5 code to make their meanings clearer. Currently broken for double precision as its kernel arg size is larger than 4096 bytes.

hummingtree committed:
Clang.

hummingtree committed:
Add fused kernel instantiations for DSLASH5_MOBIUS_PRE, M5_INV_MOBIUS, M5_INV_MOBIUS_M5_PRE.

hummingtree committed:
Push Dslash5Type to a enum class.

hummingtree committed:
Instantiate fused kernel for M5_PRE_MOBIUS_M5_INV and M5_INV_MOBIUS_M5_INV_DAG.

hummingtree committed:
Add the option store half/full vector to shared memory. Fix a few compiler warnings for enum class.

hummingtree committed:
Separate the different instantiations of the fused kernel to different files.

hummingtree committed:
Add fused kernel instantiations for asymmetric preconditioning.

hummingtree committed:
Add missing files and remove redundant files.

hummingtree committed:
Fix a bug in the code path for m5inv_m5inv, and the fused kernel now works.

hummingtree committed:
Add cmd option to toggle fused kernel at runtime. A few code cleaning.

hummingtree committed:
Abstract the function calls in dirac_mobius.cpp

hummingtree committed:
Remove dead code.

maddyscientist committed:
Merge branch 'feature/generic_kernel' of github.com:lattice/quda into feature/fused-dwf-d4-m5

maddyscientist committed:
Fused mobius kernels updated for GK.  Apply constant_kernel_arg.h header to allow for large arg size

hummingtree committed:
Remove double counting in flops for Mobius.

hummingtree committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/fused-dwf-d4-m5

hummingtree committed:
Remove the lines that modifies kernel arg.

hummingtree committed:
Apply the interior-exterior strategy for m5inv_m5inv to avoid redundant compute.

hummingtree committed:
Apply the interior-exterior strategy for m5inv to avoid redundant compute.

hummingtree committed:
Apply the interior-exterior strategy for all other types to avoid redundant compute.

hummingtree committed:
Add string to aux_base to it's much easier to know the fusion type. Fix fused kernel flops (but it's still wrong it seems)

hummingtree committed:
Merge remote-tracking branch 'origin/feature/generic_kernel' into feature/fused-dwf-d4-m5

hummingtree committed:
For fused kernel with m5inv-m5inv, revert to only using the int-ext strategy for the second m5inv.

hummingtree committed:
Fix clang unused var and clang format.

hummingtree committed:
Revert QUDA_MAX_DWF_LS back to 32.

hummingtree committed:
Revert kernel arg size limit back to 4096.

hummingtree committed:
Disable fused kernel when NVSHMEM is turned on.

hummingtree committed:
Merge remote-tracking branch 'origin/develop' into feature/fused-dwf-disable-nvshmem

hummingtree committed:
Remove dead code.

maddyscientist commented on include/dslash_helper.cuh:
```diff
@@ -4,6 +4,7 @@
 #include <gauge_field.h>
 #include <register_traits.h>
 #include <index_helper.cuh>
+#include <fast_intdiv.h>
```
This is already included a couple of lines below

hummingtree replied:
[ed49142](https://github.com/lattice/quda/pull/1163/commits/ed49142317072b3323ee86dc05f812b5efb98c74)

maddyscientist commented on include/kernels/dslash_domain_wall_4d_fused_m5.cuh:
```diff
@@ -0,0 +1,235 @@
+            out = variableInv<sync_m5inv, dagger, shared, Vector, typename Arg::Dslash5Arg>(arg, out, my_spinor_parity,
+                                                                                            0, s);
+            // Apply the m5pre.
+#if 0
```
What's the purpose of this commented out code?

hummingtree replied:
Just a reminder of the strategy that combines M5inv + M5pre that fails when kappa = 0. I will remove it.

hummingtree replied:
[ed49142](https://github.com/lattice/quda/pull/1163/commits/ed49142317072b3323ee86dc05f812b5efb98c74)

maddyscientist commented on include/dslash_quda.h:
```diff
@@ -579,6 +579,33 @@ namespace quda
                          const Complex *b_5, const Complex *c_5, const ColorSpinorField &x, int parity, bool dagger,
                          const int *comm_override, TimeProfile &profile);
 
+  void ApplyDomainWall4DM5inv(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a, double m_5,
```
Looks like clang-format hasn't been applied yet

hummingtree replied:
[835572f](https://github.com/lattice/quda/pull/1163/commits/835572fefb506c2739463ce6ded93a7ddb67437d)

maddyscientist commented on include/dslash_domain_wall_4d_fused_m5.h:
```diff
@@ -0,0 +1,153 @@
+#include <gauge_field.h>
```
Similar to dslash_coarse.hpp and coarsecoarse_op.hpp, since this file isn't a global include, and only exists to allow for fast compilation with separate compilation units, this file should be moved to `lib/dslash_domain_wall_4d_fused_m5.hpp`.

hummingtree replied:
[ed49142](https://github.com/lattice/quda/pull/1163/commits/ed49142317072b3323ee86dc05f812b5efb98c74)

maddyscientist commented on include/kernels/dslash_domain_wall_m5.cuh:
```diff
@@ -397,11 +569,13 @@ namespace quda
+        out = variableInv<sync, Arg::dagger, mobius_m5::shared()>(arg, in, parity, x_cb, s);
       } else {
-        out = constantInv<Vector>(arg, parity, x_cb, s);
+        out = constantInv<sync, Arg::dagger, mobius_m5::shared()>(arg, in, parity, x_cb, s);
```
Wondering why you made these options explicit, rather than using `Arg::dagger`  in the function?

maddyscientist replied:
Thinking about it, I guess it was to allow it being called from the fused kernels which may have multiple calls to this,, where one is daggered and one is not?

hummingtree replied:
For the m5inv_dagger * m5inv * D4 one (the second fused kernel in symmetric non-daggered) in the fused kernel, we need to first do `m5inv`, then daggered `m5inv`. So yes, the dagger-ness should/can not be inferred from the `Arg` but rather set explicitly.

hummingtree committed:
Address review comments:
- include/dslash_domain_wall_4d_fused_m5.h -> lib/dslash_domain_wall_4d_fused_m5.hpp
- Remove dead code in include/kernels/dslash_domain_wall_4d_fused_m5.cuh

hummingtree committed:
Apply clang-format.

maddyscientist commented:
Fantastic PR 😄 

Passed visual review.  Will now do some sanity checks with this before approving.

maddyscientist commented:
Ok, I've spotted one issue, which is likely an oversight.  The dslash test fails when requesting the fused operator with `--test MatPC`.

hummingtree committed:
Fix a bug where m5inv is applied after xpay (should be applied before xpay).

hummingtree commented:
@maddyscientist Good catch: this is indeed an oversight. Code has been moved around and since I did not test for `MatPC`, the issue with `Dslash4 + M5inv` was never exposed. I have fixed the issue, and `MatPC` now works for all partitions, both dagger'ed and not dagger'ed, and for both symmetric and asymmetric even-odd preconditioning.

maddyscientist reviewed (approved):
Confirmed, issue is now fixed.  Approving this PR.  Awesome work @hummingtree
